cvat-cli/src/cvat_cli/_internal/agent.py

```diff
@@ -147,39 +147,48 @@ def _validate_detection_function_compatibility(self, remote_function: dict) -> N
 
-            attrs_by_name = {attr.name: attr for attr in getattr(label, "attributes", [])}
+            for remote_sl in remote_label.get("sublabels", []):
+                sl_desc = f"sublabel {remote_sl['name']!r} of {label_desc}"
+                sl = sublabels_by_name.get(remote_sl["name"])
 
-            for remote_attr in remote_label["attributes"]:
-                attr_desc = f"attribute {remote_attr['name']!r} of {label_desc}"
-                attr = attrs_by_name.get(remote_attr["name"])
+                self._validate_sublabel_compatibility(remote_sl, sl, incompatible_msg, sl_desc)
+
+    def _validate_sublabel_compatibility(
+        self, remote_sl: dict, sl: Optional[models.Sublabel], incompatible_msg: str, sl_desc: str
+    ):
+        if not sl:
+            raise CriticalError(incompatible_msg + f"{sl_desc} is not supported.")
+
+        if remote_sl["type"] not in {"any", "unknown"} and remote_sl["type"] != sl.type:
+            raise CriticalError(
+                incompatible_msg + f"{sl_desc} has type {remote_sl['type']!r}, "
+                f"but the function object declares type {sl.type!r}."
+            )
 
-                if not attr:
-                    raise CriticalError(incompatible_msg + f"{attr_desc} is not supported.")
+        attrs_by_name = {attr.name: attr for attr in getattr(sl, "attributes", [])}
 
-                if remote_attr["input_type"] != attr.input_type.value:
-                    raise CriticalError(
-                        incompatible_msg
-                        + f"{attr_desc} has input type {remote_attr['input_type']!r},"
-                        f" but the function object declares input type {attr.input_type.value!r}."
-                    )
+        for remote_attr in remote_sl["attributes"]:
+            attr_desc = f"attribute {remote_attr['name']!r} of {sl_desc}"
+            attr = attrs_by_name.get(remote_attr["name"])
 
-                if remote_attr["values"] != attr.values:
-                    raise CriticalError(
-                        incompatible_msg + f"{attr_desc} has values {remote_attr['values']!r},"
-                        f" but the function object declares values {attr.values!r}."
-                    )
+            if not attr:
+                raise CriticalError(incompatible_msg + f"{attr_desc} is not supported.")
+
+            if remote_attr["input_type"] != attr.input_type.value:
+                raise CriticalError(
+                    incompatible_msg + f"{attr_desc} has input type {remote_attr['input_type']!r},"
+                    f" but the function object declares input type {attr.input_type.value!r}."
+                )
+
+            if remote_attr["values"] != attr.values:
+                raise CriticalError(
+                    incompatible_msg + f"{attr_desc} has values {remote_attr['values']!r},"
+                    f" but the function object declares values {attr.values!r}."
+                )
```

Comment:
Is there a way for the users to get notified that their function definition uses some unknown labels or attributes? If some annotations are silently ignored, it doesn't look really convenient to me. Maybe it's better to show some warning for ignored labels and attributes on agent startup?

Reply:
Do you mean the case where the local function has a label/sublabel/attribute that the remote function doesn't? At the moment I don't have any notifications for this case. My thinking was that since the end user only sees the remote function's labels, from their perspective everything will work as expected, even if the behind-the-scenes function object supports some extra labels.

That said, I'm not opposed to adding the warnings you propose; but I think it goes beyond the scope of this PR.